core/src/main/java/com/linecorp/armeria/common/logging/PrometheusMetricRequestDecorator.java

```diff
@@ -0,0 +1,299 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.logging;
+
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+import com.google.common.collect.ImmutableSortedMap;
+
+import com.linecorp.armeria.client.Client;
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.client.DecoratingClientFunction;
+import com.linecorp.armeria.common.Request;
+import com.linecorp.armeria.common.RequestContext;
+import com.linecorp.armeria.common.Response;
+import com.linecorp.armeria.common.RpcResponse;
+import com.linecorp.armeria.common.http.HttpSessionProtocols;
+import com.linecorp.armeria.server.DecoratingServiceFunction;
+import com.linecorp.armeria.server.Service;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+import io.prometheus.client.Collector;
+import io.prometheus.client.CollectorRegistry;
+import io.prometheus.client.Counter;
+import io.prometheus.client.Gauge;
+import io.prometheus.client.Summary;
+
+/**
+ * Provides request decorator for server and client sides. Decorators will record metrics into
+ * Prometheus {@link CollectorRegistry}.
+ * @param <I> {@link Request} type
+ * @param <O> {@link Response} type
+ */
+public final class PrometheusMetricRequestDecorator<I extends Request, O extends Response>
+        implements DecoratingServiceFunction<I, O>, DecoratingClientFunction<I, O> {
+
+    private static class MetricAdapter {
+        private final Summary timer;
+        private final Counter success;
+        private final Counter failure;
+        private final Gauge activeRequests;
+        private final Summary requestBytes;
+        private final Summary responseBytes;
+
+        MetricAdapter(PrometheusRegistryWrapper collectorRegistry,
+                      MetricLabel<?>[] metricLabels,
+                      String namePrefix,
+                      String help) {
+            final String[] metricLabelsAsString = Stream.of(metricLabels)
+                                                        .sorted()
+                                                        .map(MetricLabel::name)
+                                                        .toArray(String[]::new);
+
+            timer = create(namePrefix + "_timer",
+                           collectorRegistry,
+                           name -> addQuantiles(Summary.build())
+                                   .name(name)
+                                   .labelNames(metricLabelsAsString)
+                                   .help(help)
+                                   .create());
+            success = create(namePrefix + "_success",
```

Comment:
Should counters end with `_total`, following the guide here? https://prometheus.io/docs/practices/naming/